File: backend/models/message_model.go
package models

import (
	"chatty/backend/configs"
	"encoding/json"
	"time"
)

type Message struct {
	ID        int       `json:"id"`
	RoomID    int       `json:"roomId" validate:"required"`
	Content   string    `json:"content" validate:"required,max=500"`
	CreatedAt time.Time `json:"createdAt"`
	Username  string    `json:"username" validate:"required,max=50"`
}

type CreateMessageDTO struct {
	RoomID   int    `json:"roomId" validate:"required"`
	Content  string `json:"content" validate:"required,max=500"`
	Username string `json:"username" validate:"required,max=50"`
}

func (m *CreateMessageDTO) IsValid() (bool, error) {
	validate := configs.GetValidator()

	err := validate.Struct(m)

	if err != nil {
		return false, err
	}

	return true, nil
}

func (m *CreateMessageDTO) UnmarshalCreateMessage(data []byte) (*CreateMessageDTO, error) {
	message := &CreateMessageDTO{}
	err := json.Unmarshal(data, message)

	return message, err
}

File: talky/src/models/message.model.ts
import z from "zod";

export const messageSchema = z.object({
  content: z.string().min(1).max(500),
  roomId: z.number().int(),
  username: z.string().min(1).max(50),
  createdAt: z.coerce.date(),
  id: z.number().int(),
});

export type Message = z.infer<typeof messageSchema>;

export const createMessageSchema = messageSchema.pick({
  content: true,
  roomId: true,
  username: true,
});

export type CreateMessageDto = z.infer<typeof createMessageSchema>;
